contracts/EtomicSwap.sol

```diff
@@ -115,25 +279,143 @@ contract EtomicSwap {
 
         if (_tokenAddress == address(0)) {
-            msg.sender.transfer(_amount);
+            payable(msg.sender).transfer(_amount);
         } else {
             IERC20 token = IERC20(_tokenAddress);
             require(token.transfer(msg.sender, _amount));
         }
 
         emit SenderRefunded(_id);
     }
+
+    function senderRefundErc721(
+        bytes32 _id,
+        bytes20 _paymentHash,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _receiver
+    ) external {
+        require(payments[_id].state == PaymentState.PaymentSent);
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                _receiver,
+                msg.sender,
+                _paymentHash,
+                _tokenAddress,
+                _tokenId
+            )
+        );
+
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
+        payments[_id].state = PaymentState.SenderRefunded;
+
+        IERC721 token = IERC721(_tokenAddress);
+        token.safeTransferFrom(address(this), msg.sender, _tokenId);
+
+        emit SenderRefunded(_id);
+    }
+
+    function senderRefundErc1155(
+        bytes32 _id,
+        uint256 _amount,
+        bytes20 _paymentHash,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _receiver
+    ) external {
+        require(payments[_id].state == PaymentState.PaymentSent);
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                _receiver,
+                msg.sender,
+                _paymentHash,
+                _tokenAddress,
+                _tokenId,
+                _amount
+            )
+        );
+
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
+        payments[_id].state = PaymentState.SenderRefunded;
+
+        IERC1155 token = IERC1155(_tokenAddress);
+        token.safeTransferFrom(
+            address(this),
+            msg.sender,
+            _tokenId,
+            _amount,
+            ""
+        );
+
+        emit SenderRefunded(_id);
+    }
+
+    function onERC1155Received(
+        address, /* operator */
+        address, /* from */
+        uint256, /* id */
+        uint256, /* value */
+        bytes calldata /* data */
+    ) external pure override returns (bytes4) {
+        // Return this magic value to confirm receipt of ERC1155 token
+        return this.onERC1155Received.selector;
+    }
+
+    function onERC1155BatchReceived(
+        address, /* operator */
+        address, /* from */
+        uint256[] calldata, /* ids */
+        uint256[] calldata, /* values */
+        bytes calldata /* data */
+    ) external pure override returns (bytes4) {
+        // Return this magic value to confirm receipt of ERC1155 tokens
+        return this.onERC1155BatchReceived.selector;
```

Comment:
I think we should likely disallow receiving ERC1155 batches.

Reply:
agree, as we actually dont support batch tx now. will add `revert("Batch transfers not supported");`

Reply:
Done